typed-lambda-calculus/article.md

```diff
@@ -0,0 +1,1016 @@
+More volumes were planned originally, but after 1913 the enthusiasm seemingly ran out.
+
+It's arguably a good thing it did though, as in 1931 Kurt Gödel came out with his incompleteness theorem, proving that arithmetic can not prove its own consistency.
+This essentially has put an end to Hilbert's program.
+
+So, on some level, Russel's type theory is a bit of a failure, as it couldn't prove the consistency of mathematics.
+
+At the same time in the 1930s, as we discussed in the previous article, Alonzo Church was working on lambda calculus, also to explore the foundations of mathematics.
+
+Initial (untyped) versions of lambda calculus were susceptible to the very same (or similar) paradoxes as the naive set theories.
+So in the 1940s, in an attempt to avoid paradoxical uses of the untyped lambda calculus, Church added Russel's type theory into the mix, which got us first the simply typed lambda calculus and later other, more powerful systems.
+
+In the 1960s, Peter Landin was one of the first to notice that lambda calculus was very useful in describing semantics of programming languages.
+Arguably, this insight kickstarted the research on functional programming languages.
+
+In 1969, William Alvin Howard finalized the idea initially proposed by Haskell Curry in 1934, which equated intuitionistic logic systems (logic systems without the law of excluded middle) with typed lambda calculi.
+This kickstarted the research to find new type theories for existing logics and new logics for existing type theories.
+
+In 1972, Per Martin-Löf found a type theory corresponding to predicate logic by introducing dependent types.
+This theory became known as intuitionistic type theory or commonly Martin-Löf type theory.
+
+In 1986, Thierry Coquand and Gérard Huet created the calculus of constructions, a dependent type theory for functions.
+This theory became the basis for Coq and Lean proof assistants.
+
+Since the early 2000s, type theory research has mostly focused on the homotopy type theory (HoTT).
+In HoTT the notion of equivalence is generalized to homotopy -- i.e. the topological notion of similarity.
+HoTT turns out to be powerful enough to discuss the foundations of mathematics again and is used in the univalent foundations project, which is arguably somewhat of a spiritual successor to Hilbert's program.
+
+## Simply typed lambda calculus
+
+Untyped lambda calculus contains some "puns". In the previous article, we noted that the term $fls$ (denoting boolean zero in Church encoding) and the term $c_0$ (denoting natural zero in Church encoding) have the same representation.
+
+Furthermore, untyped lambda calculus allows for constructions like, for example, $plus\; tru\; c_2,$ which don't have a defined semantic meaning.
+
+Various typed lambda calculi introduce some restrictions on the form that terms in different expressions might have, as to avoid this punning and meaningless terms.
+
+### Lambda calculus with atomic boolean values
```

Comment:
After having read this section, I have no idea what I was supposed to take away from it. I am not a step closer to understanding typed lambda calculi, which is what the article purports to explain.